File: sdk/desktop/windows/tests/Kseal.Desktop.Tests/ConnectTrustSessionClientTests.cs
using System.Text;
using System.Text.Json;
using Xunit;

namespace Kseal.Desktop.Tests;

public class ConnectTrustSessionClientTests
{
    private static readonly TrustSessionConfig Config =
        new(new Uri("https://trust.example.com/"), "tenant-1", "app-1", Platform.Unspecified);

    [Fact]
    public void GetNonceDecodesBase64()
    {
        byte[] expected = [1, 2, 3, 4, 5];
        var transport = new FakeHttpTransport().On("GetNonce",
            FakeHttpTransport.Json($$"""{"nonce":"{{Convert.ToBase64String(expected)}}"}"""));
        var client = new ConnectTrustSessionClient(Config, transport);

        Assert.Equal(expected, client.GetNonce());
    }

    [Fact]
    public void GetNonceTargetsTrustServicePath()
    {
        var transport = new FakeHttpTransport().On("GetNonce",
            FakeHttpTransport.Json($$"""{"nonce":"{{Convert.ToBase64String(new byte[] { 9 })}}"}"""));
        new ConnectTrustSessionClient(Config, transport).GetNonce();

        Assert.EndsWith("kseal.v1.TrustService/GetNonce", transport.Requests[0].Url.AbsolutePath);
        Assert.Equal("application/json", transport.Requests[0].Headers["Content-Type"]);
        Assert.Equal("1", transport.Requests[0].Headers["Connect-Protocol-Version"]);
    }

    [Fact]
    public void GetNonceThrowsOnMissingNonce()
    {
        var transport = new FakeHttpTransport().On("GetNonce", FakeHttpTransport.Json("{}"));
        var client = new ConnectTrustSessionClient(Config, transport);
        Assert.Throws<TrustSessionException>(() => client.GetNonce());
    }

    [Fact]
    public void GetNonceThrowsOnHttpError()
    {
        var transport = new FakeHttpTransport().On("GetNonce",
            new HttpTransportResponse(500, Encoding.UTF8.GetBytes("""{"code":"internal","message":"boom"}""")));
        var client = new ConnectTrustSessionClient(Config, transport);
        var ex = Assert.Throws<TrustSessionException>(() => client.GetNonce());
        Assert.Contains("internal", ex.Message);
    }

    [Fact]
    public void VerifyAttestationParsesAcceptedSession()
    {
        string json = """
        {
          "accepted": true,
          "signedToken": "AAEC",
          "trustToken": {
            "tokenId": "tok-123",
            "expiresAt": "1700000000",
            "riskLevel": "TRUST_LEVEL_TRUSTED",
            "capabilityScope": ["read", "write"]
          }
        }
        """;
        var transport = new FakeHttpTransport().On("VerifyAttestation", FakeHttpTransport.Json(json));
        var client = new ConnectTrustSessionClient(Config, transport);

        var session = client.VerifyAttestation([1, 2, 3], 0b110, "build", "policy", "instance", [9, 9]);

        Assert.True(session.Accepted);
        Assert.Equal("tok-123", session.TokenId);
        Assert.Equal(new byte[] { 0, 1, 2 }, session.SignedToken);
        Assert.Equal(1700000000L, session.ExpiresAt);
        Assert.Equal(TrustLevel.Trusted, session.RiskLevel);
        Assert.Equal(["read", "write"], session.CapabilityScopes);
    }

    [Fact]
    public void VerifyAttestationSendsRiskBitsetAsStringAndBase64Fields()
    {
        var transport = new FakeHttpTransport().On("VerifyAttestation",
            FakeHttpTransport.Json("""{"accepted":false,"rejectionReason":"nope"}"""));
        var client = new ConnectTrustSessionClient(Config, transport);

        var session = client.VerifyAttestation([0xAA], riskBitset: 192, "b", "p", "inst", [0xBB]);

        Assert.False(session.Accepted);
        Assert.Equal("nope", session.RejectionReason);

        using var doc = JsonDocument.Parse(transport.Requests[0].Body);
        var root = doc.RootElement;
        Assert.Equal("192", root.GetProperty("riskBitset").GetString()); // int64 as string
        Assert.Equal(Convert.ToBase64String(new byte[] { 0xAA }), root.GetProperty("nonce").GetString());
        Assert.Equal(Convert.ToBase64String(new byte[] { 0xBB }), root.GetProperty("platformAttestationToken").GetString());
        Assert.Equal("PLATFORM_UNSPECIFIED", root.GetProperty("platform").GetString());
    }

    [Fact]
    public void VerifyAttestationOmitsAttestationTokenWhenEmpty()
    {
        var transport = new FakeHttpTransport().On("VerifyAttestation",
            FakeHttpTransport.Json("""{"accepted":true,"trustToken":{"tokenId":"t"}}"""));
        var client = new ConnectTrustSessionClient(Config, transport);

        client.VerifyAttestation([1], 0, "b", "p", "inst", []);

        using var doc = JsonDocument.Parse(transport.Requests[0].Body);
        Assert.False(doc.RootElement.TryGetProperty("platformAttestationToken", out _));
    }

    [Fact]
    public void ValidateRequestProofForwardsBytesAndDecodesDecision()
    {
        // RequestProofResult { decision = STEP_UP (2), reason = "mfa" }
        byte[] result = [0x08, 0x02, 0x12, 0x03, .. "mfa"u8];
        var transport = new FakeHttpTransport().On("ValidateRequestProof",
            new HttpTransportResponse(200, result));
        var client = new ConnectTrustSessionClient(Config, transport);

        byte[] proofBytes = [0xDE, 0xAD, 0xBE, 0xEF];
        var proof = new RequestProof("tok", [1], [2], 1, proofBytes);
        var decision = client.ValidateRequestProof(proof);

        Assert.Equal(TrustDecision.StepUp, decision.Decision);
        Assert.Equal("mfa", decision.Reason);
        Assert.Equal(proofBytes, transport.Requests[0].Body); // forwarded verbatim
        Assert.Equal("application/proto", transport.Requests[0].Headers["Content-Type"]);
    }

    [Fact]
    public void ValidateRequestProofThrowsOnHttpError()
    {
        var transport = new FakeHttpTransport().On("ValidateRequestProof", new HttpTransportResponse(403, []));
        var client = new ConnectTrustSessionClient(Config, transport);
        var proof = new RequestProof("tok", [1], [2], 1, [0]);
        Assert.Throws<TrustSessionException>(() => client.ValidateRequestProof(proof));
    }
}

File: sdk/desktop/macos/Tests/KsealDesktopTests/ConnectTrustSessionClientTests.swift
import XCTest
@testable import KsealDesktop

/// A scripted `HTTPTransport` that records requests and returns queued
/// responses, so the Connect JSON/proto wire mapping can be verified with no
/// sockets or server.
private final class RecordingTransport: HTTPTransport {
    struct Call {
        let url: URL
        let headers: [String: String]
        let body: Data
    }

    var responses: [HTTPResponse] = []
    private(set) var calls: [Call] = []

    func post(url: URL, headers: [String: String], body: Data) throws -> HTTPResponse {
        calls.append(Call(url: url, headers: headers, body: body))
        guard !responses.isEmpty else {
            throw TrustSessionError(message: "no scripted response")
        }
        return responses.removeFirst()
    }
}

private func json(_ object: [String: Any]) -> Data {
    try! JSONSerialization.data(withJSONObject: object)
}

final class ConnectTrustSessionClientTests: XCTestCase {

    private func makeClient(_ transport: HTTPTransport) -> ConnectTrustSessionClient {
        ConnectTrustSessionClient(
            config: TrustSessionConfig(
                baseURL: URL(string: "https://edge.kseal.test")!,
                tenantId: "tenant-1",
                appId: "com.example.app"
            ),
            transport: transport
        )
    }

    func testGetNonceParsesBase64AndHitsCorrectPath() throws {
        let transport = RecordingTransport()
        let nonce = Data([1, 2, 3, 4])
        transport.responses = [HTTPResponse(status: 200, body: json(["nonce": nonce.base64EncodedString()]))]

        let result = try makeClient(transport).getNonce()
        XCTAssertEqual(result, nonce)

        let call = try XCTUnwrap(transport.calls.first)
        XCTAssertEqual(call.url.path, "/kseal.v1.TrustService/GetNonce")
        XCTAssertEqual(call.headers["Content-Type"], "application/json")
        XCTAssertEqual(call.headers["Connect-Protocol-Version"], "1")
        let sent = try XCTUnwrap(try JSONSerialization.jsonObject(with: call.body) as? [String: Any])
        XCTAssertEqual(sent["tenantId"] as? String, "tenant-1")
        XCTAssertEqual(sent["platform"] as? String, "PLATFORM_UNSPECIFIED")
    }

    func testGetNonceRejectsEmptyNonce() {
        let transport = RecordingTransport()
        transport.responses = [HTTPResponse(status: 200, body: json(["nonce": ""]))]
        XCTAssertThrowsError(try makeClient(transport).getNonce())
    }

    func testVerifyAttestationEncodesAndDecodes() throws {
        let transport = RecordingTransport()
        transport.responses = [HTTPResponse(status: 200, body: json([
            "trustToken": [
                "tokenId": "tok-42",
                "expiresAt": "1700000999", // int64 as string
                "riskLevel": "TRUST_LEVEL_LOW_RISK",
                "capabilityScope": ["api:read", "api:write"],
            ],
            "signedToken": Data([9, 9, 9]).base64EncodedString(),
            "accepted": true,
        ]))]

        let session = try makeClient(transport).verifyAttestation(
            nonce: Data([5, 6]),
            riskBitset: 0xDEAD_BEEF,
            buildHash: "build-1",
            policyHash: "policy-1",
            instanceId: "inst-1",
            attestationToken: Data([7, 7])
        )

        XCTAssertEqual(session.tokenId, "tok-42")
        XCTAssertTrue(session.accepted)
        XCTAssertEqual(session.expiresAt, 1_700_000_999)
        XCTAssertEqual(session.riskLevel, .lowRisk)
        XCTAssertEqual(session.signedToken, Data([9, 9, 9]))
        XCTAssertEqual(session.capabilityScopes, ["api:read", "api:write"])

        let sent = try XCTUnwrap(try JSONSerialization.jsonObject(with: transport.calls[0].body) as? [String: Any])
        // 64-bit integers must be encoded as strings (proto3 JSON).
        XCTAssertEqual(sent["riskBitset"] as? String, String(UInt64(0xDEAD_BEEF)))
        XCTAssertEqual(sent["nonce"] as? String, Data([5, 6]).base64EncodedString())
        XCTAssertEqual(sent["platformAttestationToken"] as? String, Data([7, 7]).base64EncodedString())
    }

    func testVerifyAttestationDefaultsWhenFieldsOmitted() throws {
        let transport = RecordingTransport()
        // Connect omits zero-valued fields: no `accepted`, no `trustToken`.
        transport.responses = [HTTPResponse(status: 200, body: json(["rejectionReason": "blocked"]))]
        let session = try makeClient(transport).verifyAttestation(
            nonce: Data([1]), riskBitset: 0, buildHash: "", policyHash: "", instanceId: "i", attestationToken: Data()
        )
        XCTAssertFalse(session.accepted)
        XCTAssertEqual(session.rejectionReason, "blocked")
        XCTAssertEqual(session.tokenId, "")
        XCTAssertEqual(session.riskLevel, .unspecified)
    }

    func testVerifyAttestationOmitsEmptyAttestationToken() throws {
        let transport = RecordingTransport()
        transport.responses = [HTTPResponse(status: 200, body: json(["accepted": true, "trustToken": ["tokenId": "t"]]))]
        _ = try makeClient(transport).verifyAttestation(
            nonce: Data([1]), riskBitset: 0, buildHash: "", policyHash: "", instanceId: "i", attestationToken: Data()
        )
        let sent = try XCTUnwrap(try JSONSerialization.jsonObject(with: transport.calls[0].body) as? [String: Any])
        XCTAssertNil(sent["platformAttestationToken"])
    }

    func testServerErrorSurfacesConnectCode() {
        let transport = RecordingTransport()
        transport.responses = [HTTPResponse(status: 400, body: json(["code": "invalid_argument", "message": "bad nonce"]))]
        XCTAssertThrowsError(try makeClient(transport).getNonce()) { error in
            XCTAssertTrue("\(error)".contains("invalid_argument"))
        }
    }

    func testValidateRequestProofForwardsProtoAndDecodesDecision() throws {
        let transport = RecordingTransport()
        // RequestProofResult{ decision: DECISION_STEP_UP(2), reason: "mfa" }
        // field1 varint: tag 0x08, value 2; field2 string: tag 0x12, len 3, "mfa"
        let resultBytes = Data([0x08, 0x02, 0x12, 0x03, 0x6D, 0x66, 0x61])
        transport.responses = [HTTPResponse(status: 200, body: resultBytes)]

        let proof = RequestProof(
            tokenId: "tok", requestHash: Data([1]), nonce: Data([2]),
            sequence: 1, proofBytes: Data([0xAA, 0xBB])
        )
        let decision = try makeClient(transport).validateRequestProof(proof)
        XCTAssertEqual(decision.decision, .stepUp)
        XCTAssertEqual(decision.reason, "mfa")

        let call = try XCTUnwrap(transport.calls.first)
        XCTAssertEqual(call.url.path, "/kseal.v1.TrustService/ValidateRequestProof")
        XCTAssertEqual(call.headers["Content-Type"], "application/proto")
        XCTAssertEqual(call.body, Data([0xAA, 0xBB])) // proof bytes forwarded verbatim
    }
}
